docs/dev/Parsing.md

```diff
@@ -0,0 +1,103 @@
+# Parsing
+
+## Overview
+
+At its core, Wingman is a modal editor for managing planes, pilots, crew,
+flights, and locations. To achieve this modal approach,
+Wingman adopts a 2-level parsing strategy. The top level is managed inside a
+class called `WingmanParser`. Inside this class, it does two types of parsing:
+
+1. Top-level command parsing: parses the commands that do not follow any
+   logical grouping. For example, since Wingman is a `modal` editor, the
+   change of mode does not belong to any logical grouping, therefore it
+   should be parsed at the top level.
+2. Grouped command parsing: delegates the parsing of the commands that
+   belong to a certain logical grouping to the `CommandGroup` class of that
+   logical grouping.
+
+The second level of parsing happens inside the `CommandGroup` class. This
+class does a parsing of the commands similar to the 1. above. Because the
+features are similar, a base class `FactoryParser` is created to reduce code
+duplication.
+
+In designing the parsing logic, composition is preferred over inheritance,
+as inheritance makes the code less readable, increases the coupling between
+different classes, and makes it harder to test.
+
+## Parsing Logic
+
+### Step 1: Tokenization
+
+The first step in the parsing process is to "tokenize" the input string into
+a list of tokens. Here, since our parsing logic is simple, we do the
+tokenization simply by splitting the input string by white spaces.
+
+We use a `Deque` structure to store the tokens. The benefit of this is that
+we can easily pop the first token from the list, and this would take $O(1)$
+time as compared to $O(n)$ time for a `List`.
+
+The tokenization process is done in the `WingmanParser` class, inside the
+private method `tokenize`.
+
```

Comment:
thanks for the optimization, though it only has a significant effect when the command is really long..

Reply:
really really really really long lmao